gadopt/utility.py

```diff
@@ -515,3 +516,50 @@ def interpolate_1d_profile(function: Function, one_d_filename: str):
     averager = LayerAveraging(mesh, rshl if mesh.layers is None else None)
     interpolated_visc = np.interp(averager.get_layer_average(rad), rshl, one_d_data)
     averager.extrapolate_layer_average(function, interpolated_visc)
+
+
+def get_boundary_ids(mesh) -> SimpleNamespace:
+    # Firedrake adds these labels itself when creating
+    # its utility meshes.
+    if mesh.topology_dm.hasLabel("Face Sets"):
+        # This ordering is consistent among (almost) all meshes firedrake can create
+        fields = ["left", "right", "bottom", "top", "front", "back"]
+        dim = mesh.geometric_dimension()
+        if dim > 3:
+            raise ValueError(f"Cannot handle {dim} dimensional mesh")
+        # Recover the boundary ids Firedrake has inserted
+        mesh.topology_dm.markBoundaryFaces("TEMP_LABEL")
+        if mesh.topology_dm.getStratumSize("TEMP_LABEL", 1) > 0:
+            ids = {
+                mesh.topology_dm.getLabelValue("Face Sets", i)
+                for i in mesh.topology_dm.getStratumIS("TEMP_LABEL", 1).getIndices()
+            }
+        else:
+            ids = set()
+        # Not every MPI rank has every boundary of the mesh, so we to get the set of all boundaries
+        # seen by all MPI ranks
+        final_ids = set.union(*mesh.comm.allgather(ids))
+        mesh.topology_dm.removeLabel("TEMP_LABEL")
+        # Extruded standard firedrake meshes can contain numerical subdomain
+        # ids as well as "top", "bottom" labels, if the number of ids is less than
+        # the number of dimensions * 2, the string ids will always be "bottom" and "top"
+        # therefore the only time 'fields' ever appears out-of-order is if dim==3 and len(ids) < 6
+        # this implies that a 2d cartesian mesh has been extruded to a 3d cartesian mesh
+        if dim == 3 and len(final_ids) < 6:
+            kwargs = {
+                "left": 1,
+                "right": 2,
+                "front": 3,
+                "back": 4,
+                "bottom": "bottom",
+                "top": "top",
+            }
+        else:
+            kwargs = {j: i + 1 if i + 1 in final_ids else j for i, j in enumerate(fields[:2*dim])}  # noqa: E225,E226
+    # A Firedrake extruded mesh does not set boundary IDs - when this happens
+    # the boundary subdomain becomes property of the CombinedSurfaceMeasure
+    # constructed in this module, and therefore only contains "top" and
+    # "bottom" subdomains
```

Comment:
I'm not sure I understand this comment, because you are handling extruded meshes already? E.g. `mantle_convection/3d_cartesian` is an extruded `RectangleMesh`

Reply:
Yes, I think I wrote this comment before seeing the `3d_cartesian` case. Will come back to it next week.